raster-foundry/granary
# Add model routes

## Overview

This PR adds routes for `Model`s, which continue to be the worst-named datamodel in history, but we're stuck with that. It turned into a bit of a rabbit hole, getting RF deps + current http4s / doobie / etc. deps + [tapir deps](https://github.com/softwaremill/tapir/pull/235) to agree, so not everything works at the end of this. We _must_ be on circe 0.12.2+ for `circe-json-schema` to work, which we don't have a replacement available for, so we're stuck with this for a minute.

### Checklist

- [x] Any new migrations have been audited to make sure they won't kill the application while being applied
- [x] Any new API endpoints have tests

### Notes

~wip until I create the water model and add tests~

Tapir bump to 0.11.9 is to stay current so that when http4s 0.21.0 non-milestone is released and tapir for scala 2.12 includes http4s 0.21.0, we can be more confident that we can use it (smaller upgrade). Unfortunately, we can't serve the docs route right now, with the bulk of the server on http4s 0.21.0-M5, and tapir thinking that it's using 0.20.10. Presumably a non-milestone release of http4s 0.21.0 will trigger an upgrade in tapir, given [discussion](https://github.com/softwaremill/tapir/pull/235) of the motivation for the strange scala version split.

Another thing that's broken is running tests in CI. Circle CI doesn't want to run the tests from sbt with the other containers coming up, and we also didn't configure testing to be super doable from the host.

A third thing that's broken is the test database setup. I intended to spend time on that, but spent a bunch of time playing dependency jenga instead.

A fourth thing that's broken is the list endpoint maybe? It's confusing. If you hit it with a regular http request, it seems fine, but it doesn't pass tests, despite an altogether reasonable expectation I think. Anyway, I'm deferring fixing it until we have the test database setup, since that will make it a lot easier to reason about.

The schema for required arguments for the water mask model was generated using https://jsonschema.net/

There's a warning from sbt about `test in assembly := false` not doing anything, but it definitely prevents tests from running in assembly, so I'm not sure what it's on about.

## Testing Instructions

- load the database dump at `s3://rasterfoundry-development-data-us-east-1/granary/database.pgdump`
- try to list models
- it should work fine
- run tests with `./scripts/test` -- they should pass with the exception of the list endpoint

Closes #12
